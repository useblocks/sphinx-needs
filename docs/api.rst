.. _api:

Python API
==========

**Sphinx-Needs** provides an open API for other Sphinx-extensions to provide specific need-types, create needs or
make usage of the filter possibilities.

The API allows the injection of extra configuration, but
does not support manipulation of it (e.g remove need types),
to keep the final configuration transparent for the Sphinx project authors.

.. _api_configuration:

Configuration
-------------

.. automodule:: sphinx_needs.api.configuration
   :members:

.. autoclass:: sphinx_needs.functions.functions.DynamicFunction
   :members: __name__,__call__

Need
----

.. automodule:: sphinx_needs.api.need
   :members:


Exceptions
----------

.. automodule:: sphinx_needs.api.exceptions
   :members:

Data
----

.. automodule:: sphinx_needs.data
   :members: NeedsInfoType, NeedsMutable, NeedsPartType

Views
-----

These views are returned by certain functions, and injected into filters,
but should not be instantiated directly.

.. automodule:: sphinx_needs.views
   :members:
   :undoc-members:
   :special-members: __iter__, __getitem__, __len__
