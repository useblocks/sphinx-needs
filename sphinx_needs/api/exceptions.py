from sphinx.errors import SphinxError, SphinxWarning


class NeedsApiConfigException(SphinxError):
    """
    A configuration changes collides with the already provided configuration by the user.

    Example: An extension wants to add an already existing needs_type.
    """


class NeedsApiConfigWarning(SphinxWarning):
    pass


class NeedsNoIdException(SphinxError):
    pass


class NeedsDuplicatedId(SphinxError):
    pass


class NeedsStatusNotAllowed(SphinxError):
    pass


class NeedsTagNotAllowed(SphinxError):
    pass


class NeedsInvalidException(SphinxError):
    pass


class NeedsConstraintNotAllowed(SphinxError):
    pass


class NeedsConstraintFailed(SphinxError):
    pass


class NeedsInvalidOption(SphinxError):
    pass


class NeedsTemplateException(SphinxError):
    pass


class NeedsInvalidFilter(SphinxError):
    pass
