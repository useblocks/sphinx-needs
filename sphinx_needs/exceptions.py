from __future__ import annotations

from typing import Literal

from sphinx.errors import SphinxError, SphinxWarning


class NeedsApiConfigException(SphinxError):
    """
    A configuration changes collides with the already provided configuration by the user.

    Example: An extension wants to add an already existing needs_type.
    """


class InvalidNeedException(Exception):
    """Raised when a need could not be created/added, due to a validation issue."""

    def __init__(
        self,
        type_: Literal[
            "invalid_kwargs",
            "invalid_type",
            "missing_id",
            "invalid_id",
            "duplicate_id",
            "invalid_status",
            "invalid_tags",
            "invalid_constraints",
            "invalid_jinja_content",
            "invalid_template",
            "global_option",
        ],
        message: str,
    ) -> None:
        self._type = type_
        self._message = message
        super().__init__(f"{message} [{type_}]")

    @property
    def type(
        self,
    ) -> Literal[
        "invalid_kwargs",
        "invalid_type",
        "missing_id",
        "invalid_id",
        "duplicate_id",
        "invalid_status",
        "invalid_tags",
        "invalid_constraints",
        "invalid_jinja_content",
        "invalid_template",
        "global_option",
    ]:
        return self._type

    @property
    def message(self) -> str:
        return self._message


class NeedsApiConfigWarning(SphinxWarning):
    pass


class NeedsInvalidException(SphinxError):
    pass


class NeedsConstraintFailed(SphinxError):
    pass


class NeedsInvalidFilter(SphinxError):
    pass
