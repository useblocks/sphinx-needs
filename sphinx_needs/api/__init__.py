# flake8: noqa

from .configuration import (
    add_dynamic_function,
    add_extra_option,
    add_need_type,
    get_need_types,
)
from .need import add_external_need, add_need, del_need, make_hashed_id
