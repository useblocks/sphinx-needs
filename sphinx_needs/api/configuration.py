"""
API to get or add specific sphinx needs configuration parameters.

All functions here are available under ``sphinxcontrib.api``. So do not import this module directly.
"""
from typing import Any, Callable, List, Optional

from docutils.parsers.rst import directives
from sphinx.application import Sphinx
from sphinx.util.logging import SphinxLoggerAdapter

from sphinx_needs.api.exceptions import NeedsApiConfigException, NeedsApiConfigWarning
from sphinx_needs.config import NEEDS_CONFIG
from sphinx_needs.functions import register_func
from sphinx_needs.functions.functions import DynamicFunction


def get_need_types(app: Sphinx) -> List[str]:
    """
    Returns a list of directive-names from all configured need_types.

    **Usage**::

        from sphinx_needs.api import get_need_types

        all_types = get_need_types(app)

    :param app: Sphinx application object
    :return: list of strings
    """
    needs_types = app.config.needs_types
    return [x["directive"] for x in needs_types]


def add_need_type(
    app: Sphinx, directive: str, title: str, prefix: str, color: str = "#ffffff", style: str = "node"
) -> None:
    """
    Adds a new need_type to the configuration.

    The given directive must no exist, otherwise NeedsApiConfigException gets raised.

    Same impact as using :ref:`needs_types` manually.

    **Usage**::

        from sphinx_needs.api import add_need_type

        add_need_type(app, 'awesome', 'Awesome', 'AW_', '#000000', 'cloud')

    :param app: Sphinx application object
    :param directive: Name of the directive, e.g. 'story'
    :param title: Long, human-readable title, e.g. 'User-Story'
    :param prefix: Prefix, if IDs get automatically generated. E.g.: ``'US_'``
    :param color: Hex-color code used in needflow representation. Default: ``'#ffffff'``
    :param style: Plantuml-style for needflow representation. Default: 'node'
    :return: None
    """
    import sphinx_needs.directives.need

    needs_types = app.config.needs_types
    type_names = [x["directive"] for x in needs_types]

    if directive in type_names:
        raise NeedsApiConfigException(f"{directive} already exists as need type")

    needs_types.append({"directive": directive, "title": title, "prefix": prefix, "color": color, "style": style})
    app.add_directive(directive, sphinx_needs.directives.need.NeedDirective)


def add_extra_option(app: Sphinx, name: str) -> None:
    """
    Adds an extra option to the configuration. This option can then later be used inside needs or ``add_need``.

    Same impact as using :ref:`needs_extra_options` manually.

    **Usage**::

        from sphinx_needs.api import add_extra_option

        add_extra_option(app, 'my_extra_option')

    :param app: Sphinx application object
    :param name: Name as string of the extra option
    :return: None
    """

    extra_options = NEEDS_CONFIG.create_or_get("extra_options", dict)

    if name in extra_options:
        raise NeedsApiConfigWarning(f"Option {name} already registered.")

    NEEDS_CONFIG.add("extra_options", {name: directives.unchanged}, dict, append=True)


def add_dynamic_function(app: Sphinx, function: DynamicFunction, name: Optional[str] = None) -> None:
    """
    Registers a new dynamic function for sphinx-needs.

    If ``name`` is not given, the name to call the function is automatically taken from the provided function.
    The used name must be unique.

    **Usage**::

        from sphinx_needs.api import add_dynamic_function

        def my_function(app, need, needs, *args, **kwargs):
            # Do magic here
            return "some data"

        add_dynamic_function(app, my_function)

    Read :ref:`dynamic_functions` for details about how to use dynamic functions.

    :param app: Sphinx application object
    :param function: Function to register
    :param name: Name of the dynamic function as string
    :return: None
    """
    register_func(function, name)


# 'Need' is untyped, so we temporarily use 'Any' here
WarningCheck = Callable[[Any, SphinxLoggerAdapter], bool]


def add_warning(
    app: Sphinx, name: str, function: Optional[WarningCheck] = None, filter_string: Optional[str] = None
) -> None:
    """
    Registers a warning.

    A warning can be based on the result of a given filter_string or an own defined function.

    :param app: Sphinx app object
    :param name: Name as string for the warning
    :param function: function to execute to check the warning
    :param filter_string: filter_string to use for the warning
    :return: None
    """
    warnings_option = NEEDS_CONFIG.create_or_get("warnings", dict)

    if function is None and filter_string is None:
        raise NeedsApiConfigException("Function or filter_string must be given for add_warning_func")

    if function is not None and filter_string is not None:
        raise NeedsApiConfigException(
            "For add_warning_func only function or filter_string is allowed to be set, " "not both."
        )

    warning_check = function or filter_string

    if name in warnings_option:
        raise NeedsApiConfigException(f"Warning {name} already registered.")

    NEEDS_CONFIG.add("warnings", {name: warning_check}, dict, append=True)
