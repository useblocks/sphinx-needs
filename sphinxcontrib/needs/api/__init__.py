from .configuration import (  # noqa: F401
    add_dynamic_function,
    add_extra_option,
    add_need_type,
    get_need_types,
)
from .need import add_external_need, add_need, del_need, make_hashed_id  # noqa: F401
