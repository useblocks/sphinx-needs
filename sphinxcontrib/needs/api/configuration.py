"""
API to get or add specific sphinx needs configuration parameters.

All functions here are available under ``sphinxcontrib.api``. So do not import this module directly.
"""
from docutils.parsers.rst import directives
from sphinx.application import Sphinx

from sphinxcontrib.needs.api.exceptions import (
    NeedsApiConfigException,
    NeedsApiConfigWarning,
)
from sphinxcontrib.needs.config import NEEDS_CONFIG
from sphinxcontrib.needs.functions import register_func


def get_need_types(app: Sphinx):
    """
    Returns a list of directive-names from all configured need_types.

    **Usage**::

        from sphinxcontrib.needs.api import get_need_types

        all_types = get_need_types(app)

    :param app: Sphinx application object
    :return: list of strings
    """
    needs_types = app.config.needs_types
    return [x["directive"] for x in needs_types]


def add_need_type(app: Sphinx, directive, title, prefix, color="#ffffff", style="node"):
    """
    Adds a new need_type to the configuration.

    The given directive must no exist, otherwise NeedsApiConfigException gets raised.

    Same impact as using :ref:`needs_types` manually.

    **Usage**::

        from sphinxcontrib.needs.api import add_need_type

        add_need_type(app, 'awesome', 'Awesome', 'AW_', '#000000', 'cloud')

    :param app: Sphinx application object
    :param directive: Name of the directive, e.g. 'story'
    :param title: Long, human-readable title, e.g. 'User-Story'
    :param prefix: Prefix, if IDs get automatically generated. E.g.: ``'US_'``
    :param color: Hex-color code used in needflow representation. Default: ``'#ffffff'``
    :param style: Plantuml-style for needflow representation. Default: 'node'
    :return: None
    """
    import sphinxcontrib.needs.directives.need

    needs_types = app.config.needs_types
    type_names = [x["directive"] for x in needs_types]

    if directive in type_names:
        raise NeedsApiConfigException(f"{directive} already exists as need type")

    needs_types.append({"directive": directive, "title": title, "prefix": prefix, "color": color, "style": style})
    app.add_directive(directive, sphinxcontrib.needs.directives.need.NeedDirective)


def add_extra_option(app, name):
    """
    Adds an extra option to the configuration. This option can then later be used inside needs or ``add_need``.

    Same impact as using :ref:`needs_extra_options` manually.

    **Usage**::

        from sphinxcontrib.needs.api import add_extra_option

        add_extra_option(app, 'my_extra_option')

    :param app: Sphinx application object
    :param name: Name as string of the extra option
    :return: None
    """

    extra_options = NEEDS_CONFIG.create_or_get("extra_options", dict)

    if name in extra_options:
        raise NeedsApiConfigWarning(f"Option {name} already registered.")

    NEEDS_CONFIG.add("extra_options", {name: directives.unchanged}, dict, append=True)
    # extra_options[name] = directives.unchanged


def add_dynamic_function(app, function, name=None):
    """
    Registers a new dynamic function for sphinx-needs.

    If ``name`` is not given, the name to call the function is automatically taken from the provided function.
    The used name must be unique.

    **Usage**::

        from sphinxcontrib.needs.api import add_dynamic_function

        def my_function(app, need, needs, *args, **kwargs):
            # Do magic here
            return "some data"

        add_dynamic_function(app, my_function)

    Read :ref:`dynamic_functions` for details about how to use dynamic functions.

    :param app: Sphinx application object
    :param function: Function to register
    :param name: Name of the dynamic function as string
    :return: None
    """
    register_func(function, name)


def add_warning(app, name, function=None, filter_string=None):
    """
    Registers a warning.

    A warning can be based on the result of a given filter_string or an own defined function.

    :param app: Sphinx app object
    :param name: Name as string for the warning
    :param function: function to execute to check the warning
    :param filter_string: filter_string to use for the warning
    :return: None
    """
    warnings_option = NEEDS_CONFIG.create_or_get("warnings", dict)

    if function is None and filter_string is None:
        raise NeedsApiConfigException("Function or filter_string must be given for add_warning_func")

    if function is not None and filter_string is not None:
        raise NeedsApiConfigException(
            "For add_warning_func only function or filter_string is allowed to be set, " "not both."
        )

    warning_check = function or filter_string

    if name in warnings_option:
        raise NeedsApiConfigException(f"Warning {name} already registered.")

    # warnings_option[name] = warning_check
    NEEDS_CONFIG.add("warnings", {name: warning_check}, dict, append=True)
