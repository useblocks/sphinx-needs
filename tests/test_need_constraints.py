import json
import subprocess
from pathlib import Path

import pytest
from syrupy.filters import props

from sphinx_needs.api.exceptions import NeedsConstraintNotAllowed


@pytest.mark.parametrize("test_app", [{"buildername": "html", "srcdir": "doc_test/need_constraints"}], indirect=True)
def test_need_constraints(test_app, snapshot):
    app = test_app
    app.build()

    json_text = Path(app.outdir, "needs.json").read_text()
    needs_data = json.loads(json_text)
    assert needs_data == snapshot(exclude=props("created"))

    srcdir = Path(app.srcdir)
    out_dir = srcdir / "_build"

    # Check return code when "-W --keep-going" not used
    out_normal = subprocess.run(["sphinx-build", "-M", "html", srcdir, out_dir], capture_output=True)
    assert out_normal.returncode == 0

    # Check return code when only "-W" is used
    out_w = subprocess.run(["sphinx-build", "-M", "html", srcdir, out_dir, "-W"], capture_output=True)
    assert out_w.returncode >= 1

    # test if constraints_results / constraints_passed is properly set
    html = Path(app.outdir, "index.html").read_text()
    assert "<span class=\"needs_label\">constraints_results: </span>{'critical': {'check_0': False}}</span>" in html
    assert '<span class="needs_label">constraints_passed: </span>False</span>' in html

    # test force_style
    html1 = Path(app.outdir, "style_test.html").read_text()
    assert "red_bar" in html1
    assert "red_border" not in html1

    # test medium severity without force_style, appends to style
    assert "blue_border, yellow_bar" in html1


@pytest.mark.parametrize(
    "test_app", [{"buildername": "html", "srcdir": "doc_test/need_constraints_failed"}], indirect=True
)
def test_need_constraints_config(test_app):
    app = test_app

    # check that correct exception is raised
    with pytest.raises(NeedsConstraintNotAllowed):
        app.build()
