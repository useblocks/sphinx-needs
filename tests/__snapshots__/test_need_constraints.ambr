# serializer version: 1
# name: test_need_constraints[test_app0]
  dict({
    'current_version': '',
    'project': 'Python',
    'versions': dict({
      '': dict({
        'filters': dict({
        }),
        'filters_amount': 0,
        'needs': dict({
          'SECURITY_REQ': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'SECURITY_REQ',
            'created_at': '',
            'delete': None,
            'description': 'This is a requirement describing OPSEC processes.',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'test1',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'SECURITY_REQ',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'TEST DOCUMENT NEEDS CONSTRAINTS',
            'sections': list([
              'TEST DOCUMENT NEEDS CONSTRAINTS',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': None,
            'tags': list([
            ]),
            'target_id': 'SECURITY_REQ',
            'template': None,
            'title': 'test1',
            'type': 'spec',
            'type_name': 'Specification',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'SP_109F4': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
              'critical',
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
              'critical': dict({
                'check_0': True,
              }),
            }),
            'content_id': 'SP_109F4',
            'created_at': '',
            'delete': None,
            'description': 'Example of a successful constraint.',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'test2',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'SP_109F4',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
              'SECURITY_REQ',
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'TEST DOCUMENT NEEDS CONSTRAINTS',
            'sections': list([
              'TEST DOCUMENT NEEDS CONSTRAINTS',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': None,
            'tags': list([
              'critical',
            ]),
            'target_id': 'SP_109F4',
            'template': None,
            'title': 'test2',
            'type': 'spec',
            'type_name': 'Specification',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'SP_3EBFA': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
              'critical',
            ]),
            'constraints_error': 'need SP_3EBFA does not fulfill CRITICAL constraint, because tags are []',
            'constraints_passed': False,
            'constraints_results': dict({
              'critical': dict({
                'check_0': False,
              }),
            }),
            'content_id': 'SP_3EBFA',
            'created_at': '',
            'delete': None,
            'description': 'Example of a failed constraint.',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'test3',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'SP_3EBFA',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': 'debug',
            'links': list([
              'SECURITY_REQ',
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'TEST DOCUMENT NEEDS CONSTRAINTS',
            'sections': list([
              'TEST DOCUMENT NEEDS CONSTRAINTS',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': 'red_bar',
            'tags': list([
            ]),
            'target_id': 'SP_3EBFA',
            'template': None,
            'title': 'test3',
            'type': 'spec',
            'type_name': 'Specification',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'SP_CA3FB': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
              'team',
            ]),
            'constraints_passed': False,
            'constraints_results': dict({
              'team': dict({
                'check_0': False,
              }),
            }),
            'content_id': 'SP_CA3FB',
            'created_at': '',
            'delete': None,
            'description': 'Example of a failed constraint with medium severity. Note the style from :ref:`needs_constraint_failed_options`',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'FAIL_01',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'SP_CA3FB',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'TEST DOCUMENT NEEDS CONSTRAINTS',
            'sections': list([
              'TEST DOCUMENT NEEDS CONSTRAINTS',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': 'yellow_bar,',
            'tags': list([
            ]),
            'target_id': 'SP_CA3FB',
            'template': None,
            'title': 'FAIL_01',
            'type': 'spec',
            'type_name': 'Specification',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'SP_TOO_001': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'SP_TOO_001',
            'created_at': '',
            'delete': None,
            'description': 'The Tool awesome shall have a command line interface.',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'Command line interface',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'SP_TOO_001',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'TEST DOCUMENT NEEDS CONSTRAINTS',
            'sections': list([
              'TEST DOCUMENT NEEDS CONSTRAINTS',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': 'implemented',
            'style': None,
            'tags': list([
              'test',
              'test2',
            ]),
            'target_id': 'SP_TOO_001',
            'template': None,
            'title': 'Command line interface',
            'type': 'spec',
            'type_name': 'Specification',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'SP_TOO_002': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
              'critical',
            ]),
            'constraints_error': "need SP_TOO_002 does not fulfill CRITICAL constraint, because tags are ['hello', 'there']",
            'constraints_passed': False,
            'constraints_results': dict({
              'critical': dict({
                'check_0': False,
              }),
            }),
            'content_id': 'SP_TOO_002',
            'created_at': '',
            'delete': None,
            'description': 'asdf',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'CLI',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'SP_TOO_002',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'TEST DOCUMENT NEEDS CONSTRAINTS',
            'sections': list([
              'TEST DOCUMENT NEEDS CONSTRAINTS',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': 'dev',
            'style': 'red_bar',
            'tags': list([
              'hello',
              'there',
            ]),
            'target_id': 'SP_TOO_002',
            'template': None,
            'title': 'CLI',
            'type': 'spec',
            'type_name': 'Specification',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'TEST_STYLE': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
              'critical',
            ]),
            'constraints_error': 'need TEST_STYLE does not fulfill CRITICAL constraint, because tags are []',
            'constraints_passed': False,
            'constraints_results': dict({
              'critical': dict({
                'check_0': False,
              }),
            }),
            'content_id': 'TEST_STYLE',
            'created_at': '',
            'delete': None,
            'description': '',
            'docname': 'style_test',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'CLI',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'TEST_STYLE',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'TEST DOCUMENT NEEDS CONSTRAINTS',
            'sections': list([
              'TEST DOCUMENT NEEDS CONSTRAINTS',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': 'red_bar',
            'tags': list([
            ]),
            'target_id': 'TEST_STYLE',
            'template': None,
            'title': 'CLI',
            'type': 'spec',
            'type_name': 'Specification',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'TEST_STYLE2': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
              'team',
            ]),
            'constraints_passed': False,
            'constraints_results': dict({
              'team': dict({
                'check_0': False,
              }),
            }),
            'content_id': 'TEST_STYLE2',
            'created_at': '',
            'delete': None,
            'description': '',
            'docname': 'style_test',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'CLI2',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'TEST_STYLE2',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'TEST DOCUMENT NEEDS CONSTRAINTS',
            'sections': list([
              'TEST DOCUMENT NEEDS CONSTRAINTS',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': 'blue_border, yellow_bar',
            'tags': list([
            ]),
            'target_id': 'TEST_STYLE2',
            'template': None,
            'title': 'CLI2',
            'type': 'spec',
            'type_name': 'Specification',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
        }),
        'needs_amount': 8,
      }),
    }),
  })
# ---
